docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,366 @@
+down.
+
+Use-cases for Multi-Module Enclaves
+-----------------------------------
+
+There are a few implementations of multi-module enclaves based on OE SDK:
+
+- [SGX LKL](https://github.com/lsds/sgx-lkl)
+
+  This project runs a modified Linux kernel within an enclave and supports running
+  enclaves comprised of executables and shared libraries. This project supports
+  debugging of multi-module enclaves.
+
+  For reference, here is the GDB extensions module for SGX LKL.
+  `__gdb_hook_load_debug_symbols` and `LoadLibraryBreakpoint` show how
+  internal breakpoint and a "hook" (notification callback) function is used to
+  notify the debugger about libraries (modules).
+  https://github.com/lsds/sgx-lkl/blob/3038804995ff40612fda72b3f20210ceb1339ea1/tools/gdb/sgx-lkl-gdb.py#L94
+
+- [feature.openlibos](https://github.com/openenclave/openenclave/tree/feature.openlibos)
+
+  This branch of OE SDK runs a library OS within an enclave. It also supports
+  enclave comprised of multiple executables and shared libraries.
+
+  For reference, here are the additions to oegdb to support multi-module
+  debugging. Note the use of breakpoints (e.g LibraryLoadBreakpoint) and
+  notification hook functions (e.g: oe_notify_debugger_library_load)
+  https://github.com/openenclave/openenclave/blob/cf9426d83125a47c2455d3de929957f0dd73f54e/debugger/pythonExtension/gdb_sgx_plugin.py#L285-L327
+
+
+- [feature/dynamic_binding](https://github.com/openenclave/openenclave/tree/feature/dynamic_binding)
+
+  This branch of OE SDK allows loading and running multi-module enclaves with
+  the goal of permitting use of FIPS certified crypto modules. It also supports
+  debugging multi-module enclaves.
+  Here is the set of changes required to support multi-module loading.
+  It is uses a quite similar strategy to the above implementations:
+  https://github.com/openenclave/openenclave/commit/254ae94324b619861f58d0110a1affac37e5a2a4?branch=254ae94324b619861f58d0110a1affac37e5a2a4&diff=split
+
+
+The debugger changes required to support multi-module enclaves are quite similar
+in the above three implementations. The following proposal formally incorporates
+the necessary debugger changes into the Debugger Contract.
+
+Debugger Contract Extensions to Support Multi-Module Enclaves
+-------------------------------------------------------------
+
+The following `oe_debug_module_t` structure captures the information necessary
+to debug a module:
+
+```c
+
+    typedef struct _debug_module_t
+    {
+        // Magic value at start of the structure.
+        uint64_t magic;
+        // Structure version.
+        uint64_t version;
+
+        // Path of module's ELF binary.
+        char* path;
+        uint64_t path_length;
+
+        // Memory region where the enclave has been loaded.
+        uint64_t base_address;
+        uint64_t size;
+
+        // Next item in list of module.
+        struct _debug_module_t* next;
+    } oe_debug_module_t;
```

Comment:
If an oe_debug_module_t structure is added to the linked list of modules before the oe_debug_notify_module_loaded() function is called and then removed after the oe_debug_notify_module_unloaded() function is called, then a debugger that attaches to a running process could get an event on a module that it already sees as loaded, or miss an event on a module that has been unloaded.

Reply:
Thanks for pointing this out Mark. As always, I appreciate your insights.

The following three operations are themselves not atomic, and also cannot be performed together atomically:
- Check if the debugger has been attached or not
- Notify the debugger
- Add an entry to the module list

Due to the lack of atomicity, the ordering of the operations can result in different edge-cases.
(The "check if debugger has been attached or not" is ignored from the scenarios below since debugger may immediately attach/detach after the check has been performed; making the check unreliable).

In one ordering, the debugger may get double notification:
```
    1. Runtime adds module to list
        <---- Debugger attaches to the program and processes the module
    2. Runtime notifies the debugger about the module <---- Debugger will have to avoid processing the module again
    
```

In another approach, the debugger may lose track of modules
```
    1. Runtime notifies the debugger  <-- notification ignored since debugger not attached
      <------ Debugger attaches to the program, processes the list and finds no module
    2. Runtime adds the module to the list  <--- list has already been processed; module is "lost"
```

I don't know if there is a way to solve the edge-case elegantly. We could go with the first ordering above and require that the debugger be prepared for double-notifications for parsed modules upon attach.

Note: The same ordering problem exists for the list of enclaves. I'd have to dig up my notes/mails, but IIRC, this edge-case was determined not too cost-effective to address given the timing constraints of the debugger contract.

Reply:
I discussed with @auott, the strategy that was used for handling the list of enclaves what that the debugger ought to handle potential double-notification during attach.